### PR TITLE
Add option for ignoring blank lines for report line indentation

### LINKED ISSUE #13394: Add option to not announce indentation change for blank lines to Document Formatting Settings
### Is your feature request related to a problem? Please describe.

The problem is the one my [IgnoreBlanksIndentationReporting addon](https://github.com/SamKacer/IgnoreBlanksIndentation) aims to solve. It is described in the readme, but I will also rephrase it herefor full context.

Line indentation reporting is a very useful feature in NVDA, particularly for programmers since indentation is an important part of formatting code and in some languages like python it can even affect the meaning of the code. However, in its current form it can be too noisy by announcing indentation changes in contexts when it is not useful.

To explain the last point, consider this code fragment:

```
def foo():
	x = 42

	return x

def bar():
```

The current behaviour of NVDA is to report indentation changes for any line where the indentation has changed, even if the line is blank. So, the example would be read like:

```
def foo():
tab x = 42
no indent blank
tab return x
no indent blank
def bar():
```

The disadvantage for this behaviour is that for most programming languages, like python, a blank line has no semantic significance and is just used to visually separate lines of code with no change to the code's meaning. Therefore, by reporting the change of indentation upon entering a blank line and then reporting it again after landing on the next line is just noise that makes it harder to focus on understanding the code.

### Describe the solution you'd like

To increase the signal to noise ratio in situations like the one described above, it would be good if settings would have a toggle like "Announce indentation changes on blank lines". By default it would be on, since that is the current behavior. When toggled off, it would result in the above code fragment read as:

```
def foo():
tab x = 42
blank
return x

no indent def bar():
```

In this case, indentation changes are announced 2 times instead of 4 and each time it is semantically significant to understanding the code, as opposed to the default behavior.

This behavior can be implemented quite simply and I would be happy to contribute a pull request. Inside `speech.speech` the `getTextInfoSpeech` function decides whether to announce an indentation change with this `if`:

```
if reportIndentation and speakTextInfoState and allIndentation!=speakTextInfoState.indentationCache:
```

The [IgnoreBlanksIndentationAddon](https://github.com/SamKacer/IgnoreBlanksIndentation) changes this behavior so it does not announce the indentation change for a blank line by altering it to this:

```
isNotBlank = any(isinstance(t, str) and not all(c in LINE_END_CHARS for c in t) for t in textWithFields)
if reportIndentation and speakTextInfoState and isNotBlank and allIndentation!=speakTextInfoState.indentationCache:
```

Note: `LINE_END_Chars` is defined as `LINE_END_CHARS = { '\r', '\n' }`

Since the proposed solution is to implement it as a native NVDA toggleable feature, it would differ from the addon implementation by also taking into account if the setting is toggled on/off, but otherwise it wouldn't require other changes to work.

### Describe alternatives you've considered

The alternative is the aforementioned addon I developed. The major flaw with the addon implementation is that since the logic whether to announce indentation changes is buried in the `getTextInfoSpeech` function, to override this behavior, the addon must perform a monkey patch and replace `getTextInfoSpeech` with a copy where the only change is the aforementioned if statement.

The consequence of this is that any time there is any change in `getTextInfoSpeech`, no matter how minor or whether it is in any way related to indentation reporting, the addon has to be updated to also handle the new NVDA version, by defining a new version of the function with the one line change. The addon has to check the NVDA version and select the correct version of the monkey patched function. Supporting older versions of NVDA along with future releases will grow in complexity if the function is ever significantly restructured.

### PR DESCRIPTION
**Pull request by SamKacer**
### Link to issue number: 

closes #13394

### Summary of the issue:

As elaborated in #13394, line indentation reporting can be too noisy in contexts where indentation changes for blank lines aren't meaningful and distract from the actually meaningful indentation changes, such as is the case when reading source code for Python and many other programming languages.

### Description of user facing changes

A new checkbox is added to Document formatting, "Ignore blank lines for line indentation reporting". By default it is off, in which case line indentation reporting behaves as it does currently. If the user ticks the checkbox, then indentation changes are not reported for blank lines.

There is also a new unassigned global command for toggling the setting.

### Description of development approach

- new config property: `ignoreBlankLinesForRLI = boolean(default=false)`
  - short for `ignoreBlankLinesForReportLineIndentation`, which was too long and causing Flake8 issues
- new settings panel checkbox for toggling `ignoreBlankLinesForRLI`
- in `speech.speech.getTextInfoSpeech`, if `ignoreBlankLinesForRLI` is true and the line is blank, the block of code that inserts the indentation announcement and updates the indentation cache is skipped, otherwise it behaves as usual
- user guide updated with brief info on new option
- added unassigned global command for toggling `ignoreBlankLinesForRLI`
- added system test for confirming the correct behavior

### Testing strategy:

I added a new test in `symbolPronunciationTests` that confirms the correct behavior by comparing speech output when the new setting is on/off.

I also tested manually by reading the following text with "Ignore blank lines for line indentation reporting" turned off and on. In either case make sure to have line indentation reporting switched on. To quickly find the new setting, open document formatting settings with NVDA + Shift + D and press Alt + I to get to "line indentation reporting", then press tab to get to "Ignore blank lines for line indentation reporting". Alternatively, you can press alt + B twice.

#### Test text:

```
def foo():
	x = 42

	return x

def bar():
```

With current NVDA or with this change and "Ignore blank lines for line indentation reporting" turned off, the text should report 4 indentation changes like:

```
def foo():
tab x = 42
no indent blank
tab return x
no indent blank
def bar():
```

With "Ignore blank lines for line indentation reporting" turned on, it should instead report just 2 indentation changes like:

```
def foo():
tab x = 42
blank
return x

no indent def bar():
```

### Known issues with pull request:

None that I am aware of.

### Change log entries:
New features
- Add new option to document formatting settings, "Ignore blank lines for line indentation reporting" (#13394 )
Changes
Bug fixes
For Developers

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by XLTechie** [comment]
Please add an accellerator key for this checkbox. I suggest Alt+b.

**Comment by brunoprietog** [comment]
Nice! I hope this will be merged into NVDA. Just a question, instead of the indentation read change occurring on the last line, will it be possible to have NVDA say something like this?

```
def foo():
tab x = 42
blank
return x
no indent blank
def bar():
```

That makes a lot more sense to me. Thanks for bringing this to NVDA!

**Comment by XLTechie** [comment]
In testing this build, it performs in every way as I would expect.

* Tested with tones and speech (same).
* Tested switching to an editor window while on a line with an indent in that editor window.
* Tested changing to a window while on an indented line, at the same indent level as the window I departed.
* Tested moving to an indented line from a blank line, immediately after switching windows.

While I haven't reviewed the code, the behavior is good IMO, and a useful feature to have.

**Comment by XLTechie** [comment]
@brunoprietog How would it know which blanks were safe to ignore, and which blanks to reset on? I agree that your wish would make it a more useful feature, but it seems impossible to implement.

That said, if it were a combobox with options of:

* Speak normally (default)
* Ignore
* Ignore only one line

Then that third option could turn something like:

```
def foo():
	x = 42
	return x


def bar():
	pass
```

Into:

```
def foo():
tab  x = 42
return x
blank
no indent blank
def bar():
tab  pass
```

Which is more Pythonic (consistent with pep-8) for top-level functions anyway, though won't help you with methods.

**Comment by brunoprietog** [comment]
It's difficult. What occurs to me is that as the user listens to the lines, NVDA checks what is the indentation level of the next line that is not blank in the direction the user navigated (up or down).

If the user presses the up or down arrow, NVDA should check the indentation level of the next line that is not blank, in the direction in which the user navigated (up or down).
- If the indentation level is lower than the current one, on the next blank line NVDA should immediately report the indentation level change, showing that it is now lower, even if it is on a blank line.
- If the indentation level is higher than the current one, NVDA should notify the user of the indentation change to the higher level when the user reaches that line.

Here is an example:

```python
def foo
	pass



def bar
	pass

def baz


	pass
```

Reading it from top to bottom, NVDA should say this:

```
def foo
tab pass
no indent blank
blank
blank
def bar
tab pass
no indent blank
def baz
blank
blank
tab pass
```

Reading it from the bottom up, NVDA should say this:

```
tab pass
no indent blank
blank
def baz
blank
tab pass
no indent def bar
blank
blank
blank
tab pass
no indent def foo
```

That's more intuitive for me, at least in my case

**Comment by cary-rowen** [comment]
Hi @SamKacer 

This is a nice change and I like it.
I believe this will be beneficial to all, Whether this option can be checked by default.

Thanks

**Comment by XLTechie** [comment]
NVDA has never tried to be predictive about what navigation the user might do 
next. I understand why you want this, and it even seems reasonable. But that is 
a more controversial change than the original request I think.

I would rather see this go forward as is (or with my one-line-only 
modification), then see the proposal to be predictive as a feature request after 
this feature is established.

All kinds of things open up if we're willing to predict what the user might want 
to know or do next, but I think that will need deeper discussion than is called 
for by this simple change. A proposal like that has potentially concerning 
side-effects, and will need exploration.

**Comment by leonardder** [comment]
How ought this to behave when indentation reporting with tones is on?

**Comment by SamKacer** [comment]
@XLTechie wrote
> Please add an accelerator key for this checkbox. I suggest Alt+b.
> 
> Also, would an unassigned gesture to toggle it be useful?

Thanks, will do those when I have time. This is my first time adding any new config property, so please let me know about anything I missed. Actually, how do I add the accelerator key? I haven't noticed that when making my changes. Also in which file are the commands for toggling setting properties defined?

to do:

- [ ] add accelerator key for new setting
- [ ] add unassigned toggle command for new setting

**Comment by XLTechie** [comment]
> how do I add the accelerator key? I haven't noticed that when making my changes.

Ampersand before the letter in the setting name. So if it was "Announce blanks", 
and you wanted the key to be alt+b, you would write it as "Announce &blanks".

Btw, while you're in gui/settingsDialogs.py, please consider fixing the spelling 
of "panel" on lines 2393 and 2400.

> Also in which file are the commands for toggling setting properties defined?

globalCommands.py

See script_toggleProgressBarOutput for an example.
Also see script_reportLinkDestinationInWindow, for an example of proper typing, 
which isn't done in the other one.

**Comment by XLTechie** [comment]
> How ought this to behave when indentation reporting with tones is on?

Based on testing it, it behaves exactly as I expected.

There are no de-indent beeps on the lines that are blank.

**Comment by SamKacer** [comment]
In regards to @brunoprietog request. He has already asked this in #13394 and I gave a very long reply there [in this comment](https://github.com/nvaccess/nvda/issues/13394#issuecomment-1059757162). Linking in case someone wants full context.

Thanks for the suggestion, but this would require a significant overhaul of the indentation reporting code. I am not entirely sure if I would even prefer that behavior. I think I might prefer if the rules for indentation reporting are kept simpler without special cases like this. In any case, I'm not sure the amount of effort to implement would be that worth it.

**Comment by SamKacer** [comment]
> How ought this to behave when indentation reporting with tones is on?

@leonardder the behavior should be the same whether you are using tones or speech

**Comment by SamKacer** [comment]
Ok, this is officially ready for review.

I implemented the 2 features @XLTechie suggested. I added a system test. I fixed all the Flake8 issues. The summary is up to date.

Please let me know if anything else is missing or if anything could be improved. Thanks

File: user_docs/en/userGuide.t2t
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:2123]: Fix typo

```suggestion
This may be useful when reading a document where blank lines are used to separate indented blocks of text, such as in programming source code.
```

File: tests/system/robot/symbolPronunciationTests.py
Comment (seanbudd) [inline tests/system/robot/symbolPronunciationTests.py:547]: Now that this is more complex, can this get a sensible default too to match the default in configSpec, `SymLevel.SOME`

```suggestion
		symbolLevel: SymLevel: SymLevel.SOME,
```

File: source/globalCommands.py
Comment (seanbudd) [inline source/globalCommands.py:630]: Fixup CamelCase for `toggleIgnore`

```suggestion
	def script_toggleIgnoreBlankLinesForReportLineIndentation(self, gesture: inputCore.InputGesture) -> None:
```

File: tests/system/robot/symbolPronunciationTests.py
Comment (seanbudd) [inline tests/system/robot/symbolPronunciationTests.py:567]: same here re-sensible defaults

File: tests/system/robot/symbolPronunciationTests.py
Comment (SamKacer) [reply tests/system/robot/symbolPronunciationTests.py:547]: I am just wondering if this adds value here. Adding defaults for the 2 new properties was helpful because that way I didn't need to make as many changes. Every place that uses `_doTest` or `_setConfig` already specifies symbol level. Also although the default in `configSpec` for sym level is some, this value isn't actually used in any of these tests. So isn't adding a default for that param unnecessary and if it was to be added, wouldn't NONE or ALL be a more useful default value for these tests?

File: source/speech/speech.py
Comment (SamKacer) [reply source/speech/speech.py:1550]: Isn't this a case perhaps where having this expression on one line makes things more readable? The preceding line succintly summarizes what that expression computes. I am imagining in most cases when someone is reading this part of the code they just want to understand what is happening at a high level. Spreading out this expression over multiple lines, particularly since it computes a simple and very straightforward thing means a reader has to arrow through more lines of code. When it is on one line, they can skip past it with one key stroke if they don't care about the details (probably 90% of the time) and if they do care about the details then they can move by word.

Alternatively, what if this expression was moved into a helper function where the expression was spread out more as you suggest, but that way this bit of code where it would be called would remain succinct?

File: tests/system/robot/symbolPronunciationTests.py
Comment (seanbudd) [reply tests/system/robot/symbolPronunciationTests.py:547]: I think it would be useful to have here in-case future tests do not wish to change the default value for symbol level, and instead wish to test the default symbol level.
Previously the function was `_setSymbolLevel` which was for explicitly changing the symbol from default. Now that it is `_setConfig` and empty call to `_setConfig` should not change the default config.

File: source/speech/speech.py
Comment (seanbudd) [reply source/speech/speech.py:1550]: a helper function may be better here - but the current line is not parseable for me at all. I had to separate it to newlines understand the change here.

File: tests/system/robot/symbolPronunciationTests.py
Comment (SamKacer) [reply tests/system/robot/symbolPronunciationTests.py:547]: I think that definitely makes sense. I am just concerned that this could be overengineering, since this is a change to support a testing pattern that isn't being used and might never be.

Consider that if I accepted this code change, the subsequent build would fail. Beyond the syntax being incorrect, SymLevel doesn't actually have a `SOME` member. It just has `NONE` and `ALL`, because that is the only options used in these tests. So if I was to make this change, I would also have to add that member. But since I would be adding it just in case someone in the future wanted to use it, then wouldn't it make sense to also add `MOST` in case someone wanted to use that in the future. And then it would make most sense to fill in the other enums like `IndentationReporting` with their missing options.

Instead wouldn't it be better if whoever would have such a use case in the future to be the one to add these things and make the changes that are necessary? I think that would avoid making changes and additions that might not even end up being used.

Comment (seanbudd) [reply tests/system/robot/symbolPronunciationTests.py:547]: I'd rather not accept a misleading function into the code base. I think the change of purpose to `_setConfig` warrants the change. The only additional required code involves adding a `SOME` member.

Comment (SamKacer) [reply tests/system/robot/symbolPronunciationTests.py:547]: But isn't it misleading to add the `SOME` member if it isn't actually being used in any test? I am not sure if I understand how `_setConfig` is now misleading. Back when it was `_setSymbolLevel` it didn't have any dfault and it wasn't considered misleading then. If the name of the function is what is misleading, then what if it was changed to something like `_changeConfig` or `_overrideConfig`?

@michaelDCurran @Qchristensen could we please get a third opinion on this thread?

Comment (SamKacer) [reply tests/system/robot/symbolPronunciationTests.py:547]: Alternatively, another option could be to return `_setSymbolLevel` to what it was and add a `_setIndentationReporting` function.

Comment (michaelDCurran) [reply tests/system/robot/symbolPronunciationTests.py:547]: Since we are dealing with configuration values here, I think they should either all be explicit, or all be optional with defaults.
As you don't wish to repeat indentation values that won't change much, I think the only option then is to make everything optional as @seanbudd  said.
However, I also like the alternative of adding _setIndentationReporting and keeping these functions separate.
So in short, if you are unwilling to make the symbol level value optional, then perhaps go with _setIndentationReporting.
However, as @seanbudd  is the NvDA core developer reviewing this pr, he should have the final say.

Comment (SamKacer) [reply tests/system/robot/symbolPronunciationTests.py:547]: @michaelDCurran thanks for your input

@seanbudd would you kindly make the code change you had in mind?

Comment (seanbudd) [reply tests/system/robot/symbolPronunciationTests.py]: That's been done

**Review by seanbudd — APPROVED** [review]

**Review by Qchristensen — APPROVED** [review]
User guide reads well, thanks!

**Pull request merged**
